File: q0207/q0207.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        preReqMap = {}
        for i in range(numCourses):
            preReqMap[i] = []
        for course, preReqCourse in prerequisites:
            preReqMap[course].append(preReqCourse)

        visitedCourse = set()

        def dfsCheckCycle(course):
            if course in visitedCourse:
                return False
            if preReqMap[course] == []:
                return True

            visitedCourse.add(course)
            for preReq in preReqMap[course]:
                if not dfsCheckCycle(preReq):
                    return False
            visitedCourse.remove(course)
            preReqMap[course] = []
            return True

        for course in range(numCourses):
            if not dfsCheckCycle(course):
                return False
        return True

File: q0207/q0207.java
import java.util.*;

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        Map<Integer, List<Integer>> preReqMap = new HashMap<>();
        for (int i = 0; i < numCourses; i++) {
            preReqMap.put(i, new ArrayList<Integer>());
        }
        for (int[] pair : prerequisites) {
            int course = pair[0];
            int preReq = pair[1];
            preReqMap.get(course).add(preReq);

        }
        HashSet<Integer> visited = new HashSet<>();
        for (int i = 0; i < numCourses; i++) {
            if (!dfsCheckCycle(visited, preReqMap, i))
                return false;
        }
        return true;
    }

    private boolean dfsCheckCycle(HashSet<Integer> visited,
            Map<Integer, List<Integer>> preReqMap,
            int c) {
        if (visited.contains(c))
            return false;
        if (preReqMap.get(c).size() == 0)
            return true;
        visited.add(c);
        for (int preReq : preReqMap.get(c)) {
            if (!dfsCheckCycle(visited, preReqMap, preReq))
                return false;
        }
        visited.remove(c);
        preReqMap.get(c).clear();
        return true;
    }
}